xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/hibernate/query/HqlQueryExecutor.java

```diff
@@ -143,17 +159,46 @@ protected static boolean isSafeSelect(String statementString)
         return HqlQueryUtils.isSafe(completeStatement);
     }
 
+    private boolean isSafe(String statement) throws QueryException
+    {
+        // An empty statement is safe
+        if (statement.isEmpty()) {
+            return true;
+        }
+
+        // HqlQueryUtils#isSafe only works with complete statements
+        String completeStatement = toCompleteShortForm(statement);
+
+        // Parse and validate the statement
+        ComponentManager componentManager = this.componentManagerProvider.get();
+        try {
+            for (HQLQueryValidator validator : componentManager
+                .<HQLQueryValidator>getInstanceList(HQLQueryValidator.class)) {
+                Optional<Boolean> result = validator.isSafe(completeStatement);
+                if (result.isPresent()) {
+                    return result.get();
+                }
+            }
+        } catch (ComponentLookupException e) {
+            throw new QueryException("Failed to get query statement validators",
+                new DefaultQuery(statement, Query.HQL, this), e);
+        }
+
+        // Assume the statement is safe if no validator says otherwise
+        return true;
```

Comment:
Wouldn't it be safer to do this the other way around (i.e., return `false`), to ensure that if an attacker manages to unregister a component for whatever reason, the attacker can't execute arbitrary queries? In particular, given that the default implementation always returns a non-empty `Optional` and thus this code should never be reached.

Reply:
I hesitated a lot. It's safer to return false, but I was afraid that we would end up blocking a lot of stuff and making the debug a lot harder because getting there means that there is a serious problem with the validators.

But you are right, I did not though about the un registration of the default validator as an attack. I guess another approach could be to `@Inject` the default validator and call it as last resort, even if it's already supposed to be called above. It means that the executor init will fail if there is a problem with the default validator, which also probably make more sense.

Reply:
Yes, indeed, having a proper error when the default validator cannot be used would be great. Btw., it would be really cool to have this method as an (internal) API to re-use in other parts of XS. I'm currently working on some code that would need this.

Reply:
>  Btw., it would be really cool to have this method as an (internal) API to re-use in other parts of XS. I'm currently working on some code that would need this.

OK, I will refactor that a bit.

Reply:
> I guess it's a good way to fix my current ClassFanOutComplexity.

Looks like it did not help that much in the end.

Anyway now you can simply use `HQLStatementValidator#isSafe` to validate a HQL statement.